receiver/kubeletstatsreceiver/kubelet/client.go

```diff
@@ -0,0 +1,205 @@
+type ClientConfig struct {
+	// AuthType should be "tls" or "serviceAccount". TLS requires that the cert/key
+	// paths below also be set.
+	AuthType AuthType `mapstructure:"auth_type"`
+	// CacertPath should be the path to the root CAs. Used to verify the kubelet
+	// server's certificates.
+	CacertPath string `mapstructure:"ca_cert_path"`
+	// ClientKeyPath should be the path to the TLS client cert. Presented to the
+	// other side of the kubelet connection.
+	ClientCertPath string `mapstructure:"client_cert_path"`
+	// ClientKeyPath should be the path to the TLS client key. Must correspond to
+	// the ClientCertPath, making a key pair to present to the kubelet.
+	ClientKeyPath string `mapstructure:"client_key_path"`
+	// InsecureSkipVerify controls whether the client verifies the server's
+	// certificate chain and host name.
+	InsecureSkipVerify bool `mapstructure:"insecure_skip_verify"`
+}
+
+// A kubelet client, which mostly handles auth when talking to a kubelet server.
+// Marshaling/unmarshaling should be performed by the caller.
+type Client struct {
+	c       *http.Client
+	baseURL string
+	token   string
+	logger  *zap.Logger
+}
+
+func NewClient(baseURL string, cfg *ClientConfig, logger *zap.Logger) (*Client, error) {
+	var client *Client
+	var err error
+	switch cfg.AuthType {
+	case AuthTypeTLS:
+		client, err = tlsClient(baseURL, cfg)
+	case AuthTypeServiceAccount:
+		client, err = serviceAccountClient(baseURL)
+	default:
+		return nil, errors.New("cfg.AuthType not found")
+	}
+	if err != nil {
+		return nil, err
+	}
+	if baseURL == "" {
+		// this requires HostNetwork to be turned on
+		hostname, err := os.Hostname()
+		if err != nil {
+			return nil, err
+		}
+		client.baseURL = fmt.Sprintf("http://%s:10250", hostname)
+	} else {
+		client.baseURL = baseURL
+	}
+	client.logger = logger
+	return client, nil
+}
+
+func tlsClient(baseURL string, cfg *ClientConfig) (*Client, error) {
+	rootCAs, err := systemCertPoolPlusPath(cfg.CacertPath)
+	if err != nil {
+		return nil, err
+	}
+
+	clientCert, err := tls.LoadX509KeyPair(cfg.ClientCertPath, cfg.ClientKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	tr := http.DefaultTransport.(*http.Transport).Clone()
+	tr.TLSClientConfig = &tls.Config{
+		RootCAs:            rootCAs,
+		Certificates:       []tls.Certificate{clientCert},
+		InsecureSkipVerify: cfg.InsecureSkipVerify,
+	}
+
+	return &Client{
+		c:       &http.Client{Transport: tr},
+		baseURL: baseURL,
+	}, nil
+}
+
+func serviceAccountClient(baseURL string) (*Client, error) {
+	const svcAcctTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	token, err := ioutil.ReadFile(svcAcctTokenPath)
+	if err != nil {
+		return nil, err
+	}
+	const caCertPath = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+	rootCAs, err := systemCertPoolPlusPath(caCertPath)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		c: &http.Client{
+			Transport: &http.Transport{
```

Comment:
Should clone default?

Reply:
Ah yes, I updated the first client but not this one. Will fix.